clang/lib/Sema/SemaExpr.cpp

```diff
@@ -229,6 +229,21 @@ bool Sema::DiagnoseUseOfDecl(NamedDecl *D, ArrayRef<SourceLocation> Locs,
                !checkAllowedSYCLInitializer(VD, /*CheckValueDependent =*/true))
         SYCLDiagIfDeviceCode(*Locs.begin(), diag::err_sycl_restrict)
             << Sema::KernelConstStaticVariable;
+      // Emit an error if a const global array's size exceeds SPIR-V
+      // intermediate format limitations for SPIRV compatible devices.
+      else if (IsConst && VD->hasGlobalStorage() &&
+               Context.getTargetInfo().getTriple().isSPIR() &&
+               !isArrayZeroInitialized(VD, /*CheckValueDependent =*/true)) {
+        QualType Ty = VD->getType();
+        if (const auto *CAType = dyn_cast<ConstantArrayType>(Ty)) {
```

Comment:
I think you need to use `getAsConstantArrayType()` here instead of `dyn_cast`. In another PR, we ran into issues with arrays not being recognized correctly with `dyn_cast` when it was an elaborated type. @erichkeane can you confirm?

Reply:
Yep, this could very well be a typedef wrapping the constant array type.